oracle.go

```diff
@@ -0,0 +1,336 @@
+			logger.Warningf("cleanup of %q failed: %v", name, err)
+		}
+	}
+}
+
+// CompletedTokens looks at the list of tokens and finds what referenced txns
+// are completed, and then returns the set of tokens that are completed.
+func (o *DBOracle) CompletedTokens(tokens []string) (map[string]bool, error) {
+	objectIds := make([]bson.ObjectId, 0, len(tokens))
+
+	// The nonce is generated during preparing, and if 2 flushers race,
+	// only one nonce makes it into the final transaction. However, other
+	// nonces can also be considered 'completed'. (afaict, they are ignored,
+	// thus won't be applied and can be considered completed.)
+	for _, token := range tokens {
+		objId := txnTokenToId(token)
+		objectIds = append(objectIds, objId)
+	}
+	query := o.working.Find(bson.M{"_id": bson.M{"$in": objectIds}})
+	query = query.Select(bson.M{"_id": 1})
+	iter := query.Iter()
+	var txnDoc struct {
+		Id bson.ObjectId `bson:"_id"`
+	}
+	foundIdHex := make(map[string]bool, len(objectIds))
+	for iter.Next(&txnDoc) {
+		foundIdHex[txnDoc.Id.Hex()] = true
+	}
+	if err := iter.Close(); err != nil {
+		if err != mgo.ErrNotFound {
+			// Not found is ok, the transactions may not be complete
+			return nil, err
+		}
+	}
+	result := make(map[string]bool, len(foundIdHex))
+	// because multiple tokens could map to a single txn, we iterate the
+	// passed in tokens instead of caching them in the map.
+	for _, token := range tokens {
+		objIdHex := txnTokenToId(token).Hex()
+		if foundIdHex[objIdHex] {
+			result[token] = true
+		}
+	}
+	o.checkedTokens += uint64(len(tokens))
+	o.completedTokens += uint64(len(result))
+	o.foundTxns += uint64(len(foundIdHex))
+	return result, nil
+}
+
+// RemoveTxns can be used to flag that a given transaction should not
+// be considered part of the valid set.
+func (o *DBOracle) RemoveTxns(txnIds []bson.ObjectId) error {
+	_, err := o.working.RemoveAll(bson.M{"_id": bson.M{"$in": txnIds}})
+	if err != nil {
+		return fmt.Errorf("error removing transaction ids: %v", err)
+	}
+	return nil
+}
+
+type dbIterWrapper struct {
+	iter *mgo.Iter
+}
+
+var _ OracleIterator = (*dbIterWrapper)(nil)
+
+var EOF = fmt.Errorf("end of transaction ids")
+
+func (d *dbIterWrapper) Next() (bson.ObjectId, error) {
+	var txnId struct {
+		Id bson.ObjectId `bson:"_id"`
+	}
+	if d.iter.Next(&txnId) {
+		return txnId.Id, nil
+	}
+	if err := d.iter.Close(); err != nil {
+		return txnId.Id, err
+	}
+	return txnId.Id, EOF
+}
+func (d *dbIterWrapper) Close() error {
+	return d.iter.Close()
+}
+
+// IterTxns lets you iterate over all of the transactions that have
+// not been removed.
+func (o *DBOracle) IterTxns() (OracleIterator, error) {
+	iter := o.working.Find(nil).Select(bson.M{"_id": 1}).Iter()
+	return &dbIterWrapper{iter: iter}, nil
+}
+
+// MemOracle uses a temporary table on disk to track what transactions are
+// considered completed and purgeable.
+type MemOracle struct {
+	txns            *mgo.Collection
+	completed       map[bson.ObjectId]bool
```

Comment:
So this still has the excessive memory consumption problem ... I think that's ok for this PR but I do wonder if we should read the iterator and feed that into a temporary collection or an on disk data structure to avoid the issue of mgopurge dying when pruning large Juju 1.25 databases.